oracle_fdw.c

```diff
@@ -728,6 +730,69 @@ _PG_init(void)
 	on_proc_exit(&exitHook, PointerGetDatum(NULL));
 }
 
+/*
+ * Deparse LIMIT clause translated for ORACLE into FETCH FIRST N ROWS ONLY
+ * If OFFSET is set then do not append clause at all, the limit +offset
+ * will be applied localy.
+ */
+char *
+deparseLimit(PlannerInfo *root, struct OracleFdwState *fdwState, RelOptInfo *baserel)
+{
+	StringInfoData limit_clause;
+
+	initStringInfo(&limit_clause);
+
+	if (root->parse->limitCount != NULL && root->parse->limitOffset == NULL)
+	{
+		if (IsA(root->parse->limitCount, Const) && !((Const *) root->parse->limitCount)->constisnull)
+		{
+			int limit_at_end = false;
+
+			/*
+			 * we must be sure that there is no ORDER BY clause that
+			 * appears after the LIMIT clause otherwise we will have
+			 * wrong results.
+			 */
```

Comment:
First, it seems like you are doing it the other way around: you don't push down a `LIMIT` that is *after* an `ORDER BY`. So the comment should be "we must make sure that there is no ORDER BY before the LIMIT", right?

Second, what is the reason for that? If we push down the `ORDER BY`, why shouldn`t we push down the `LIMIT` too?
I must be missing something, but I don't see how that could produce wrong results.
Are you thinking of differences in the collation?

Reply:
About the ORDER BY restriction this is because the clause might not return the same order from Oracle than from PostgreSQL. This restriction is wrong for numeric column but not for string column where the sort order can be different. I've taken the choice of the restriction but this could be assume and we can push the LIMIT clause in any cases or control this with an option.

About the dbserver value in the regression test, yes sorry I forgot to remove it before pushing. I will fix that as well as the typo.

Concerning `FETCH FIRST n ROWS ONLY` for Oracle version < 12.2 I can change the clause to use ROWNUM just like I'm doing the the revert transformation from Ora2Pg. I agree about cursor and useless of LIMIT but if I recall well oracle_fdw prefetch the first 100 rows, and limit is often used to just return the first row and as you notice this could also help the Oracle planner to choose a better plan.

Reply:
Searching for a variable to change the oracle_fdw behavior following the Oracle version I see that there is none. Does it make sens for you if I add a new foreign server option **dbversion** to allow the user to set the major version of the Oracle server? It can also be auto-detected at remote connection time but this will add the execution of an extra remote query. Or I can do both, setting the foreign server option will suppress the remote call.

Reply:
There is the `oracleServerVersion` function that I use for `oracle_diag`.
I think it would be OK to call that in `oracleGetSession` immediately after a connection is established, construct a number like PostgreSQL's `server_version_num` from it and store it in the `oracleSession`.
The overhead would be small, and connections are kept open anyway.
Then one can check the server version whenever necessary.

I understand your concern about `ORDER BY`, but then a `LIMIT` without an `ORDER BY` is mostly useless.
[The code](https://github.com/laurenz/oracle_fdw/blob/620495448681ea3314dd1b9660dc013707d2bfb8/oracle_fdw.c#L843) suggests that `ORDER BY` on string columns is not pushed down anyway (for the same reasons), so I think you don't have to worry about that.

Reply:
Ah right, I've not paid attention to the ORDER BY clause restriction. So that's fine I will apply the LIMIT clause anyway. Yes I have seen the call to oracleServerVersion in the diagnostic, I will reuse it. I will work on it later today.